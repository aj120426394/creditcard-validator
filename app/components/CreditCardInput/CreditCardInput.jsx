// @flow
import * as React from 'react';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCcVisa, faCcMastercard, faCcAmex, faCcDiscover } from '@fortawesome/free-brands-svg-icons';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import { zDepth1 } from '../../styled-utils/shadow';

const OuterContainer = styled.div`
  ${zDepth1};
  width: 100%;
  height: 40px;
  padding: 3px;
  box-sizing: border-box;
  border-radius: 5px;
  position: relative;
  background-image: ${(props) => {
    if (props.valid === 'valid') {
      return 'linear-gradient(to right, #3ec8ac 0%, #4e90a4 100%), linear-gradient(to right, #3ec8ac 0%, #4e90a4 100%)';
    } else if (props.valid === 'invalid') {
      return 'linear-gradient(to right, #f44336 0%, #e91e63 100%), linear-gradient(to right, #f44336 0%, #e91e63 100%)';
    } else {
      return 'linear-gradient(to right, #c73ec8 0%, #75faf6 100%), linear-gradient(to right, #c73ec8 0%, #75faf6 100%)';
    }
  }};
`;

const InnerContainer = styled.div`
  background: #fff;
  box-sizing: border-box;
  border-radius: 5px;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  overflow: hidden;
  justify-content: space-around;
  padding: 0 10px;
  position: relative;
`;
const IconContainer = styled.div`
  width: 5%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  color: grey;
  z-index: 2;
  position: relative;
`;

const CaptionContainer = styled.div`
  position: absolute;
  bottom: 0;
  transform: translateY(100%);
  font-size: 0.8em;
  padding-left: 10px;
  box-sizing: border-box;
  color: red;
`;

export const InputBox = styled.input`
  position: relative;
  z-index: 2;
  border: none;
  width: 100%;
  height: 100%;
  background: transparent;
  appearance: textfield;
  font-size: 18px;
  &:focus {
    outline: none;
  }
  &::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    -webkit-appearance: none;
  }
  &::placeholder {
    font-size: 18px;
  }
`;
const Placeholder = styled.input`
  z-index: 1;
  position: absolute;
  width: 90%;
  height: 100%;
  left: 9.8px;
  box-sizing: border-box;
  border: none;
  font-size: 18px;

  &::placeholder {
    font-size: 18px;
  }
`;
type Valid = 'valid' | 'invalid' | 'matching';

type Props = {
  getValidResult?: (Valid) => void
};

type State = {
  icon: typeof faCcVisa,
  placeholder: string,
  shownValue: string,
  valid: Valid
};

export default class CreditCardInput extends React.Component<Props, State> {
  state = {
    icon: null,
    placeholder: 'Number on your credit card',
    shownValue: '',
    valid: 'matching'
  };
  input: ?HTMLInputElement;

  /**
   * Check for the card type and number of digits
   * @param {string} cardNum - card number
   * @return {[string, number]} - card type and number of digits
   */
  checkCreditCardTypeAndNumberOfDigits = (cardNum: string): [string, number] => {
    const regexAmex = /^(34|37)/;
    const regexDiscover = /^6011/;
    const regexMastercard = /^5[1-5]/;
    const regexVisa = /^4/;
    let numberOfDigits = 0;

    if (regexAmex.exec(cardNum) !== null) {
      // AMEX
      numberOfDigits = 15;
      return ['amex', numberOfDigits];
    } else if (regexDiscover.exec(cardNum) !== null) {
      // Discover
      numberOfDigits = 16;
      return ['discover', numberOfDigits];
    } else if (regexMastercard.exec(cardNum) !== null) {
      // Master
      numberOfDigits = 16;
      return ['master', numberOfDigits];
    } else if (regexVisa.exec(cardNum) !== null) {
      // Visa
      numberOfDigits = cardNum.length < 14 ? 13 : 16;
      return ['visa', numberOfDigits];
    } else {
      numberOfDigits = 0;
      return ['none', numberOfDigits];
    }
  };

  /**
   * Validate the card number.
   * @param {string} cardNum - User's input card number
   * @param {number} numberOfDigits - The number of the digits regards to card type.
   * @returns {string} - 'valid' | 'invalid' | 'matching'
   */
  cardValidation = (cardNum: string, numberOfDigits: number): Valid => {
    const startWith = /^[3456]/;
    // If first digits do not match the requirement
    if (cardNum.length !== 0 && startWith.exec(cardNum) === null) {
      return 'invalid';
    }
    // If user input more than 4 digits and there is no card type match
    if (cardNum.length >= 4 && numberOfDigits === 0) {
      return 'invalid';
    }

    if (cardNum.length !== numberOfDigits || cardNum.length === 0) {
      return 'matching';
    }
    return this.luhnMethod(cardNum) ? 'valid' : 'invalid';
  };

  /**
   * Using Luhn algorithm to validate the number.
   * @param {string} cardNum - User's input card number
   * @returns {boolean} - True if the card number is validated.
   */
  luhnMethod = (cardNum: string): boolean => {
    const validArray = [];
    let validResult = 0;
    for (let i = 0; i < cardNum.length; i++) {
      const index = cardNum.length - 1 - i;
      if ((i + 1) % 2 === 0) {
        validArray[index] = 2 * Number(cardNum.charAt(index));
      } else {
        validArray[index] = Number(cardNum.charAt(index));
      }
    }
    validArray.forEach((i) => {
      if (i >= 10) {
        validResult = validResult + (i % 10);
        validResult = validResult + 1;
      } else {
        validResult = validResult + i;
      }
    });
    return validResult % 10 === 0;
  };

  /**
   * Forming the input card number in the right format regards to the card type.
   * @param {string} cardNum - User's input card number.
   * @param {string} cardType - User's card type.
   * @returns {string} - Formatted card number.
   */
  formCardNumberToTemplate = (cardNum: string, cardType: string): string => {
    switch (cardType) {
      case 'amex':
        // 0000 000000 00000
        return `${cardNum.substring(0, 4)} ${cardNum.substring(4, 10)} ${cardNum.substring(10)}`.trim();
      case 'discover':
        // 0000 0000 0000 0000
        return `${cardNum.substring(0, 4)} ${cardNum.substring(4, 8)} ${cardNum.substring(8, 12)} ${cardNum.substring(
          12
        )}`.trim();
      case 'master':
        // 0000 0000 0000 0000
        return `${cardNum.substring(0, 4)} ${cardNum.substring(4, 8)} ${cardNum.substring(8, 12)} ${cardNum.substring(
          12
        )}`.trim();
      case 'visa':
        // 0000 0000 0000 0000
        // 0000 0000 0000 0
        return `${cardNum.substring(0, 4)} ${cardNum.substring(4, 8)} ${cardNum.substring(8, 12)} ${cardNum.substring(
          12
        )}`.trim();
      default:
        return cardNum;
    }
  };

  onInputChange = (e: SyntheticEvent<HTMLInputElement>) => {
    let input = e.target.value.replace(/\s/g, '').trim();
    const [cardType, numberOfDigits] = this.checkCreditCardTypeAndNumberOfDigits(input);
    let placeholder = input.slice(0);
    let shownValue = input.slice(0);
    while (placeholder.length < numberOfDigits) {
      placeholder = placeholder + '0';
    }

    const valid = this.cardValidation(input, numberOfDigits);
    let icon = null;
    if (valid === 'invalid') {
      icon = faTimesCircle;
    } else {
      switch (cardType) {
        case 'amex':
          icon = faCcAmex;
          break;
        case 'discover':
          icon = faCcDiscover;
          break;
        case 'master':
          icon = faCcMastercard;
          break;
        case 'visa':
          icon = faCcVisa;
          break;
        default:
          icon = null;
      }
    }

    this.setState(
      (prevState) => {
        if (numberOfDigits === 0 || input.length <= numberOfDigits) {
          placeholder =
            input.length === 0 ? 'Number on your credit card' : this.formCardNumberToTemplate(placeholder, cardType);
          return {
            placeholder: placeholder,
            shownValue: this.formCardNumberToTemplate(shownValue, cardType),
            valid: valid,
            icon: icon
          };
        } else {
          return {};
        }
      },
      () => {
        if (this.props.getValidResult) {
          this.props.getValidResult(this.state.valid);
        }
      }
    );
  };

  onClickOnReset = () => {
    if (this.state.valid === 'invalid') {
      this.setState(
        {
          shownValue: '',
          placeholder: 'Number on your credit card',
          valid: 'matching',
          icon: null
        },
        () => {
          if (this.input) {
            this.input.focus();
          }
        }
      );
    }
  };

  render() {
    return (
      <OuterContainer valid={this.state.valid}>
        <InnerContainer>
          <InputBox
            innerRef={(ref) => {
              this.input = ref;
            }}
            type="tel"
            maxlength="5"
            value={this.state.shownValue}
            onChange={this.onInputChange}
          />
          <Placeholder placeholder={this.state.placeholder} />
          <IconContainer>
            {this.state.icon && (
              <FontAwesomeIcon
                icon={this.state.icon}
                size="lg"
                style={{ cursor: this.state.valid === 'invalid' ? 'pointer' : 'default' }}
                onClick={this.onClickOnReset}
              />
            )}
          </IconContainer>
        </InnerContainer>
        {this.state.valid === 'invalid' && <CaptionContainer>Invalid card number</CaptionContainer>}
      </OuterContainer>
    );
  }
}
