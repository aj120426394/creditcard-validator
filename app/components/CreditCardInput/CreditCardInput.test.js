// @flow
import * as React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCcVisa, faCcMastercard, faCcAmex, faCcDiscover } from '@fortawesome/free-brands-svg-icons';
import { faTimesCircle } from '@fortawesome/free-solid-svg-icons';
import CreditCardInput, { InputBox } from './CreditCardInput';

describe('CreditCardInput: ', () => {
  const mountWrapper = mount(<CreditCardInput />);

  test('render: Snapshot test', () => {
    expect(toJson(mountWrapper)).toMatchSnapshot();
  });

  test('card validator', () => {
    const validator = mountWrapper.instance().cardValidation;
    expect(validator('4111111111111111', 16)).toBe('valid');
    expect(validator('4111111111111', 13)).toBe('invalid');
    expect(validator('4012888888881881', 16)).toBe('valid');
    expect(validator('378282246310005', 15)).toBe('valid');
    expect(validator('6011111111111117', 16)).toBe('valid');
    expect(validator('60185', 0)).toBe('invalid');
    expect(validator('5105105105105100', 16)).toBe('valid');
    expect(validator('5105105105105106', 16)).toBe('invalid');
    expect(validator('9111111111111111', 0)).toBe('invalid');
    expect(validator('40523', 13)).toBe('matching');
  });

  test('card type checker', () => {
    const checker = mountWrapper.instance().checkCreditCardTypeAndNumberOfDigits;
    expect(checker('4111111111111111')).toEqual(['visa', 16]);
    expect(checker('4111111111111')).toEqual(['visa', 13]);
    expect(checker('4012888888881881')).toEqual(['visa', 16]);
    expect(checker('378282246310005')).toEqual(['amex', 15]);
    expect(checker('6011111111111117')).toEqual(['discover', 16]);
    expect(checker('5105105105105100')).toEqual(['master', 16]);
    expect(checker('5105105105105106')).toEqual(['master', 16]);
    expect(checker('9111111111111111')).toEqual(['none', 0]);
    expect(checker('40523')).toEqual(['visa', 13]);
  });

  test('luhn algorithm', () => {
    const luhnMethod = mountWrapper.instance().luhnMethod;
    expect(luhnMethod('5105105105105100')).toBe(true);
    expect(luhnMethod('5105105105105106')).toBe(false);
  });

  test('format card number into template', () => {
    const formCardNumberToTemplate = mountWrapper.instance().formCardNumberToTemplate;
    expect(formCardNumberToTemplate('4111111111111111', 'visa')).toBe('4111 1111 1111 1111');
    expect(formCardNumberToTemplate('4111111111111', 'visa')).toBe('4111 1111 1111 1');
    expect(formCardNumberToTemplate('378282246310005', 'amex')).toBe('3782 822463 10005');
    expect(formCardNumberToTemplate('6011111111111117', 'discover')).toBe('6011 1111 1111 1117');
    expect(formCardNumberToTemplate('5105105105105100', 'master')).toBe('5105 1051 0510 5100');
    expect(formCardNumberToTemplate('9111111111111111', 'none')).toBe('9111111111111111');
  });

  describe('Interaction: User ', () => {
    const inputBox = mountWrapper.find(InputBox);

    test('input number that is Visa card', () => {
      const inputValue = '41111111111';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('4111 1111 1110 0');
      expect(mountWrapper.state('shownValue')).toBe('4111 1111 111');
      expect(mountWrapper.state('icon')).toBe(faCcVisa);

      const inputValue2 = '4111111111111111';
      inputBox.simulate('change', { target: { value: inputValue2 } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('4111 1111 1111 1111');
      expect(mountWrapper.state('shownValue')).toBe('4111 1111 1111 1111');
      expect(mountWrapper.state('icon')).toBe(faCcVisa);


      const inputValue3 = '4111111111111112';
      inputBox.simulate('change', { target: { value: inputValue3 } });
      expect(mountWrapper.state('valid')).toBe('invalid');
      expect(mountWrapper.state('placeholder')).toBe('4111 1111 1111 1112');
      expect(mountWrapper.state('shownValue')).toBe('4111 1111 1111 1112');
      expect(mountWrapper.state('icon')).toBe(faTimesCircle);

    });

    test('input number that is Amex card', () => {
      const inputValue = '37828224631';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('3782 822463 10000');
      expect(mountWrapper.state('shownValue')).toBe('3782 822463 1');
      expect(mountWrapper.state('icon')).toBe(faCcAmex);


      const inputValue2 = '378282246310005';
      inputBox.simulate('change', { target: { value: inputValue2 } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('3782 822463 10005');
      expect(mountWrapper.state('shownValue')).toBe('3782 822463 10005');
      expect(mountWrapper.state('icon')).toBe(faCcAmex);

      const inputValue3 = '378282246310008';
      inputBox.simulate('change', { target: { value: inputValue3 } });
      expect(mountWrapper.state('valid')).toBe('invalid');
      expect(mountWrapper.state('placeholder')).toBe('3782 822463 10008');
      expect(mountWrapper.state('shownValue')).toBe('3782 822463 10008');
      expect(mountWrapper.state('icon')).toBe(faTimesCircle);

    });

    test('input number that is Master card', () => {
      const inputValue = '51051051051';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('5105 1051 0510 0000');
      expect(mountWrapper.state('shownValue')).toBe('5105 1051 051');
      expect(mountWrapper.state('icon')).toBe(faCcMastercard);

      const inputValue2 = '5105105105105100';
      inputBox.simulate('change', { target: { value: inputValue2 } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('5105 1051 0510 5100');
      expect(mountWrapper.state('shownValue')).toBe('5105 1051 0510 5100');
      expect(mountWrapper.state('icon')).toBe(faCcMastercard);

      const inputValue3 = '5105105105105108';
      inputBox.simulate('change', { target: { value: inputValue3 } });
      expect(mountWrapper.state('valid')).toBe('invalid');
      expect(mountWrapper.state('placeholder')).toBe('5105 1051 0510 5108');
      expect(mountWrapper.state('shownValue')).toBe('5105 1051 0510 5108');
      expect(mountWrapper.state('icon')).toBe(faTimesCircle);

    });

    test('input number that is Discover card', () => {
      const inputValue = '60111111111';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1110 0000');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 111');
      expect(mountWrapper.state('icon')).toBe(faCcDiscover);

      const inputValue2 = '6011111111111117';
      inputBox.simulate('change', { target: { value: inputValue2 } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('icon')).toBe(faCcDiscover);

      const inputValue3 = '6011111111111112';
      inputBox.simulate('change', { target: { value: inputValue3 } });
      expect(mountWrapper.state('valid')).toBe('invalid');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1111 1112');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 1111 1112');
      expect(mountWrapper.state('icon')).toBe(faTimesCircle);

    });

    test('delete the number in the input box', () => {
      inputBox.simulate('change', { target: { value: '' } });
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('Number on your credit card');
      expect(mountWrapper.state('shownValue')).toBe('');
      expect(mountWrapper.state('icon')).toBe(null);
    });

    test('input more the the required number of digits', () => {
      const inputValue = '6011111111111117';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('icon')).toBe(faCcDiscover);

      const inputValue2 = '60111111111111178';
      inputBox.simulate('change', { target: { value: inputValue2 } });
      expect(mountWrapper.state('valid')).toBe('valid');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 1111 1117');
      expect(mountWrapper.state('icon')).toBe(faCcDiscover);
    });

    test('click on reset button', () => {
      const inputValue = '6011111111111112';
      inputBox.simulate('change', { target: { value: inputValue } });
      expect(mountWrapper.state('valid')).toBe('invalid');
      expect(mountWrapper.state('placeholder')).toBe('6011 1111 1111 1112');
      expect(mountWrapper.state('shownValue')).toBe('6011 1111 1111 1112');
      expect(mountWrapper.state('icon')).toBe(faTimesCircle);

      const resetButton = mountWrapper.find(FontAwesomeIcon);
      resetButton.simulate('click');
      expect(mountWrapper.state('valid')).toBe('matching');
      expect(mountWrapper.state('placeholder')).toBe('Number on your credit card');
      expect(mountWrapper.state('shownValue')).toBe('');
      expect(mountWrapper.state('icon')).toBe(null);
    })
  });

});
