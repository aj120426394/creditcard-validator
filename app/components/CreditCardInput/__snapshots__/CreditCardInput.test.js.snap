// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCardInput:  render: Snapshot test 1`] = `
<CreditCardInput>
  <CreditCardInput__OuterContainer
    valid="matching"
  >
    <div
      className="CreditCardInput__OuterContainer-npp3l3-0 ixexGs"
    >
      <CreditCardInput__InnerContainer>
        <div
          className="CreditCardInput__InnerContainer-npp3l3-1 cPBAnQ"
        >
          <CreditCardInput__InputBox
            innerRef={[Function]}
            maxlength="5"
            onChange={[Function]}
            type="tel"
            value=""
          >
            <input
              className="CreditCardInput__InputBox-npp3l3-3 bnQDvB"
              onChange={[Function]}
              type="tel"
              value=""
            />
          </CreditCardInput__InputBox>
          <CreditCardInput__Placeholder
            placeholder="Number on your credit card"
          >
            <input
              className="CreditCardInput__Placeholder-npp3l3-4 eVDvks"
              placeholder="Number on your credit card"
            />
          </CreditCardInput__Placeholder>
          <CreditCardInput__IconContainer>
            <div
              className="CreditCardInput__IconContainer-npp3l3-2 bdFWjF"
            />
          </CreditCardInput__IconContainer>
        </div>
      </CreditCardInput__InnerContainer>
    </div>
  </CreditCardInput__OuterContainer>
</CreditCardInput>
`;
